[clang-tidy] Add modernize-use-std-format check

mikecrowe opened the pull request:
Add a new clang-tidy check that converts absl::StrFormat (and similar functions) to std::format (and similar functions.)

Split the configuration of FormatStringConverter out to a separate Configuration class so that we don't risk confusion by passing two boolean configuration parameters into the constructor. Add AllowTrailingNewlineRemoval option since we never want to remove trailing newlines in this check.

mikecrowe committed:
[clang-tidy] Add modernize-use-std-format check

Add a new clang-tidy check that converts absl::StrFormat (and similar
functions) to std::format (and similar functions.)

Split the configuration of FormatStringConverter out to a separate
Configuration class so that we don't risk confusion by passing two
boolean configuration parameters into the constructor. Add
AllowTrailingNewlineRemoval option since we never want to remove
trailing newlines in this check.

Differential Revision: https://reviews.llvm.org/D154287

mikecrowe commented:
This check was originally proposed and reviewed at https://reviews.llvm.org/D154287 last year but never landed since I did not (and still do not) have commit access. Near the end of that review, @PiotrZSL wrote:
> To be honest, I do not see to much use case for this check. But that's fine.

which may mean that it should not land. I've created this pull request for discoverability by those looking for such a check since Phabricator is no longer used.

EugeneZelenko commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.h:
```diff
@@ -0,0 +1,50 @@
+#ifndef LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MODERNIZE_USESTDFORMATCHECK_H
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MODERNIZE_USESTDFORMATCHECK_H
+
+#include "../ClangTidyCheck.h"
```
Included twice.

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Remove duplicate inclusion of ClangTidyCheck.h.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -0,0 +1,108 @@
+
+  DiagnosticBuilder Diag =
+      diag(StrFormatCall->getBeginLoc(), "use '%0' instead of %1")
+      << ReplacementFormatFunction << OldFunction->getIdentifier();
```
`<< OldFunction; ` should be sufficient.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -0,0 +1,108 @@
+      diag(StrFormatCall->getBeginLoc(), "use '%0' instead of %1")
+      << ReplacementFormatFunction << OldFunction->getIdentifier();
+  Diag << FixItHint::CreateReplacement(
+      CharSourceRange::getTokenRange(StrFormatCall->getBeginLoc(),
```
getBeginLoc/getEndLoc could be changed into [getSourceRange](https://clang.llvm.org/doxygen/classclang_1_1Stmt.html#a6844392ee6148b5fe821f211b95e5d1b) ()

PiotrZSL commented on clang-tools-extra/clang-tidy/utils/FormatStringConverter.cpp:
```diff
@@ -627,7 +628,8 @@ void FormatStringConverter::finalizeFormatText() {
   // It's clearer to convert printf("Hello\r\n"); to std::print("Hello\r\n")
   // than to std::println("Hello\r");
-  if (StringRef(StandardFormatString).ends_with("\\n") &&
+  if (Config.AllowTrailingNewlineRemoval &&
+      StringRef(StandardFormatString).ends_with("\\n") &&
       !StringRef(StandardFormatString).ends_with("\\\\n") &&
       !StringRef(StandardFormatString).ends_with("\\r\\n")) {
```
Avoid constructing StringRef multiple times.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.h:
```diff
@@ -0,0 +1,48 @@
+
+namespace clang::tidy::modernize {
+
+/// Convert calls to absl::StrFormat-like functions to std::format.
```
synchronize with release notes

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:
```diff
@@ -0,0 +1,97 @@
+  // CHECK-FIXES: std::format("c_str() removal iterator width only:{:{}} width and precision:{:{}.{}} precision only:{:.{}}", *is1, 3, *is2, 4, 2, *is3, 5);
+
+  return s1 + s2 + s3 + s4 + s5 + s6;
+}
```
add few tests with macros. Mainly when part of call is behind macro (name, arguments).

mikecrowe replied:
I've added some tests. The check as currently implemented sees through and replaces the macro for the function call and argument but does not replace macros in the argument list.

mikecrowe replied:
https://github.com/llvm/llvm-project/blob/84a81343cdda86efa34a9340f2a7ab7d2b388731/clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp#L99

@PiotrZSL, were these the kinds of tests you were expecting when you made your request and are you happy with the behaviour of the check for them? If not, please can you provide more specific advice?

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:
```diff
@@ -0,0 +1,84 @@
+
+  to ensure that the output will continue to be the unsigned representation
+  of -42 and the signed representation of 0xffffffff (often 4294967254
+  and -1 respectively.) When `false` (which is the default), these casts
```
`.)` -> `).`

PiotrZSL reviewed (approved):
To be honest I'm not sure why it wasn't merged, there were some dependences on other changes and modernize-use-std-print check, and somehow it fall under radar.

Overall looks fine. Please fix pointed out nits that pop up.

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Use getSourceRange rather than getBeginLoc()/getEndLoc().

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Synchronise description of check in UseStdFormatCheck.h with the
more-detailed description in the release notes.

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

.) -> ).

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Add some tests that include macro expansions.

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Avoid calling StringRef constructor multiple times.

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Run clang-format

SimplyDanny commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -0,0 +1,107 @@
+  const Expr *StrFormatCall = StrFormat->getCallee();
+  if (!Converter.canApply()) {
+    DiagnosticBuilder Diag = diag(StrFormat->getBeginLoc(),
+                                  "unable to use '%0' instead of %1 because %2")
```
```suggestion
                                  "unable to use '%0' instead of '%1' because %2")
```

mikecrowe replied:
The use of `OldFunction->getIdentifier()` means that the string passed for `%1` is already surrounded by quotes. You can see this in the lit test strings:
```
// CHECK-MESSAGES: [[@LINE-1]]:13: warning: use 'std::format' instead of 'StrFormat' [modernize-use-std-format]
```

mikecrowe replied:
Yes. I can't say that I particularly like the situation and if I should be using a different method to get the function name then I'd be happy to change it. If someone changes that behaviour in the future then the lit tests will all start failing so they will be alerted to the need to do something.

SimplyDanny commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  DiagnosticBuilder Diag =
+      diag(StrFormatCall->getBeginLoc(), "use '%0' instead of %1")
```
```suggestion
      diag(StrFormatCall->getBeginLoc(), "use '%0' instead of '%1'")
```

SimplyDanny commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:
```diff
@@ -0,0 +1,84 @@
+removing now-unnecessary calls to ``std::string::c_str()`` and
+``std::string::data()``.
+
+In other words, it turns lines like:
```
```suggestion
For example, it turns lines like
```

SimplyDanny commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:
```diff
@@ -0,0 +1,84 @@
+   and ``ssize_t``) as the opposite signedness to ensure that the output
+   would matches that of a simple wrapper for ``std::sprintf`` that
+   accepted a C-style variable argument list. For example, with
+   `StrictMode` enabled:
```
```suggestion
   `StrictMode` enabled,
```

SimplyDanny commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:
```diff
@@ -0,0 +1,84 @@
+    unsigned int u = 0xffffffff;
+    return strprintf("%d %u\n", i, u);
+
+  would be converted to:
```
```suggestion
  would be converted to
```

SimplyDanny reviewed (approved):
Only a few nits concerning non-functional code/text. Other than that LGTM.

mikecrowe committed:
Update clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst

Co-authored-by: Danny Mösch <danny.moesch@icloud.com>

mikecrowe committed:
Update clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst

Co-authored-by: Danny Mösch <danny.moesch@icloud.com>

mikecrowe committed:
Update clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst

Co-authored-by: Danny Mösch <danny.moesch@icloud.com>

PiotrZSL reviewed (approved):
LGTM

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:
```diff
@@ -0,0 +1,120 @@
+  // The function call is replaced even though it comes from a macro.
+#define FORMAT absl::StrFormat
+  auto s1 = FORMAT("Hello %d", 42);
+  // CHECK-MESSAGES: [[@LINE-1]]:13: warning: use 'std::format' instead of 'StrFormat' [modernize-use-std-format]
+  // CHECK-FIXES: std::format("Hello {}", 42);
```
but, this shound't be fixed, so for macros like this there should be no fixes

mikecrowe replied:
Would you be willing to let this change land as it is and then I'll work on fixing this problem in a subsequent change for both this check and for _modernise-use-std-print_ that suffers from the same problem?

PiotrZSL replied:
Thing is, that macro can be configuration dependent, and removing macros is not a good idea.
Add also test where everything is under macro (if there arent anny).

Once build will pass, I will land this change, and fixes could be done in separate PR for both checks.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format.cpp:
```diff
@@ -0,0 +1,120 @@
+  // behaviour is required so that that <inttypes.h> macros are replaced.
+#define FORMAT_STRING "Hello %s"
+  auto s2 = absl::StrFormat(FORMAT_STRING, 42);
+  // CHECK-MESSAGES: [[@LINE-1]]:13: warning: use 'std::format' instead of 'StrFormat' [modernize-use-std-format]
+  // CHECK-FIXES: std::format("Hello {}", 42);
```
no fix also should be here

mikecrowe replied:
I think this is required behaviour otherwise the check wouldn't be able to convert the unfortunately normal:
```C++
std::string f(int64_t i, uint64_t u) {
  return strprintf("%" PRId64 ", %" PRIu64, i, u);
}
```
to the expected
```C++
std::string f(int64_t i, uint64_t u) {
  return std::format("{}, {}", i, u);
}
```
The _modernise-use-std-print_ check currently does the same thing and is [documented](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-std-print.html) as doing so.

mikecrowe commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:
```diff
@@ -0,0 +1,84 @@
+
+.. code-block:: c++
+
+  return absl::StrFormat("The %s is %3d\n", description.c_str(), value);
```
The newline in the format string here should be removed since it isn't present in the format string on line 22 below.

mikecrowe committed:
fixup! [clang-tidy] Add modernize-use-std-format check

Remove incorrect newline that won't be removed from the format string
during conversion.

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-format.rst:
```diff
@@ -0,0 +1,84 @@
+
+.. option:: StrFormatLikeFunctions
+
+   A semicolon-separated list of (fully qualified) extra function names to
```
The `extra` is wrong here. If you don't pass an option the default is described later, but when you do pass something, then default `absl::StrFormat` is not part of the checked function names.

mikecrowe replied:
True. I'll fix that.

mikecrowe replied:
The fix for use-std-print is in https://github.com/llvm/llvm-project/pull/91069 .

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-format-custom.cpp:
Should the check options use the simpler format? E.g. 
```
CheckOptions: 
  modernize-use-std-format.StrFormatLikeFunctions: 'unqualified_strprintf;::strprintf; mynamespace::strprintf2'
```

mikecrowe replied:
Thanks! I wasn't aware that was possible since I started with all these checks before fbf611ed2a768999202e2c5e1e1a6c3c6bb94725 and had been battling with the JSON syntax.

5chmidti reviewed (approved):
Just nits, otherwise LGTM!

mikecrowe committed:
Remove unnecessary variable as suggested by @5chmidti

mikecrowe commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -0,0 +1,107 @@
+                                         getLangOpts());
+  const Expr *StrFormatCall = StrFormat->getCallee();
+  if (!Converter.canApply()) {
+    DiagnosticBuilder Diag = diag(StrFormat->getBeginLoc(),
```
True. I'll fix that.

mikecrowe commented on clang-tools-extra/clang-tidy/modernize/UseStdFormatCheck.cpp:
```diff
@@ -0,0 +1,107 @@
+                                  "unable to use '%0' instead of %1 because %2")
+                             << ReplacementFormatFunction
+                             << OldFunction->getIdentifier()
+                             << Converter.conversionNotPossibleReason();
```
OK. It looks like use-std-print needs that fix too, so I'll do that in a separate change. Is there a way to get the lit test to check for this?

mikecrowe replied:
The use-std-print equivalent is in https://github.com/llvm/llvm-project/pull/91071 .

mikecrowe committed:
Fix mistake in docs pointed out by @5chmidti

mikecrowe committed:
use-std-format: Use simpler CheckOptions syntax in lit tests

As suggested by @5chmidti.

mikecrowe committed:
use-std-format: Ensure that function name gets underlined in diagnostic

as suggested by @5chmidti.

mikecrowe commented:
@PiotrZSL wrote:
> Once build will pass, I will land this change, and fixes could be done in separate PR for both checks.

The build seems to be passing. Please can you land this if you are happy?

Thanks.

kadircet commented:
hi! this seems to be crashing for certain code patterns, https://github.com/llvm/llvm-project/issues/92896 has a minimal reproducer
